src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,790 @@
+            >>> gf(-1).is_one()
+            False
+            >>> gf(1).is_one()
+            True
+        """
+        return 1 == fq_default_is_one(self.val, self.ctx.val)
+
+    def __richcmp__(self, other, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fq_default cannot be ordered")
+
+        if not typecheck(other, fq_default):
+            other = self.ctx.any_as_fq_default(other)
+
+        if typecheck(other, fq_default):
+            res = (self.ctx == (<fq_default>other).ctx) and \
+                  fq_default_equal(self.val, (<fq_default>other).val, self.ctx.val)
+            if op == 2:
+                return res
+            else:
+                return not res
+        else:
+            return NotImplemented
+
+    # =================================================
+    # Generic arithmetic required by flint_scalar
+    # =================================================
+
+    def _neg_(self):
+        cdef fq_default res
+        res = self.ctx.new_ctype_fq_default()
+        fq_default_neg(res.val, self.val, self.ctx.val)
+        return res
+
+    def _add_(self, other):
+        other = self.ctx.any_as_fq_default(other)
+        if other is NotImplemented:
+            return NotImplemented
+
+        cdef fq_default res
+        res = self.ctx.new_ctype_fq_default()
+        fq_default_add(res.val, self.val, (<fq_default>other).val, self.ctx.val)
+        return res
+
+    @staticmethod
+    def _sub_(left, right):
+        # Case when left and right are already fq_default
+        if typecheck(left, fq_default) and typecheck(right, fq_default):
+            if not (<fq_default>left).ctx == (<fq_default>right).ctx:
+                raise ValueError("moduli must match")
+
+        # Otherwise, exactly one of these values is fq_default, if it's left
+        # convert the right
+        elif typecheck(left, fq_default):
+            right = (<fq_default>left).ctx.any_as_fq_default(right)
+            if right is NotImplemented:
+                return NotImplemented
+        # Otherwise convert the left
+        else:
+            left = (<fq_default>right).ctx.any_as_fq_default(left)
+            if left is NotImplemented:
+                return NotImplemented
+
+        # Now perform subtraction
+        cdef fq_default res
+        res = (<fq_default>left).ctx.new_ctype_fq_default()
+        fq_default_sub(res.val, (<fq_default>left).val, (<fq_default>right).val, res.ctx.val)
+        return res
+
+    def _mul_(self, other):
+        other = self.ctx.any_as_fq_default(other)
+        if other is NotImplemented:
+            return NotImplemented
+
+        cdef fq_default res
+        res = self.ctx.new_ctype_fq_default()
+        fq_default_mul(res.val, self.val, (<fq_default>other).val, self.ctx.val)
+        return res
+
+    @staticmethod
+    def _div_(left, right):
+        # Case when left and right are already fq_default
```

Comment:
Having `__rdiv__` call here reversing the order of the arguments places us in a somewhat awkward position where we don't know which argument is actually an `fq_default`.

Reply:
i tried to fix this with a `swap=False` bool:

```py
    def __rsub__(self, other):
        other = self._any_as_self(other)
        if other is NotImplemented:
            return NotImplemented
        return self._sub_(other, swap=True)
```